libraries/AP_CANManager/AP_CANManager.h

```diff
@@ -0,0 +1,151 @@
+ *
+ * This file is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Code by Siddharth Bharat Purohit
+ */
+
+#pragma once
+
+#include <AP_HAL/AP_HAL.h>
+
+#if HAL_MAX_CAN_PROTOCOL_DRIVERS
+
+#include <AP_Param/AP_Param.h>
+#include "AP_SLCANIface.h"
+#include "AP_CANDriver.h"
+
+class AP_CANManager
+{
+public:
+    AP_CANManager();
+
+    /* Do not allow copies */
+    AP_CANManager(const AP_CANManager &other) = delete;
+    AP_CANManager &operator=(const AP_CANManager&) = delete;
+
+    static AP_CANManager* get_singleton()
+    {
+        return _singleton;
+    }
+
+    enum LogLevel {
+        LOG_NONE,
+        LOG_ERROR,
+        LOG_WARNING,
+        LOG_INFO,
+        LOG_DEBUG,
+    };
+
+    enum Driver_Type : uint8_t {
+        Driver_Type_None = 0,
+        Driver_Type_UAVCAN = 1,
+        Driver_Type_KDECAN = 2,
+        Driver_Type_ToshibaCAN = 3,
+        Driver_Type_PiccoloCAN = 4,
+        Driver_Type_TestCAN = 5,
+    };
+
+    void init(void);
+
+    // returns number of active CAN Drivers
+    uint8_t get_num_drivers(void) const
+    {
+        return _num_drivers;
+    }
+
+    // return driver for index i
+    AP_CANDriver* get_driver(uint8_t i) const
+    {
+        if (i < HAL_NUM_CAN_IFACES) {
+            return _drivers[i];
+        }
+        return nullptr;
+    }
+
+    // Method to log status and debug information for review while debugging
+    void log_text(AP_CANManager::LogLevel loglevel, const char *tag, const char *fmt, ...);
+    
+    uint32_t log_retrieve(char* data, uint32_t max_size) const;
+
+    // return driver type index i
+    Driver_Type get_driver_type(uint8_t i) const
+    {
+        if (i < HAL_NUM_CAN_IFACES) {
+            return _driver_type_cache[i];
+        }
+        return Driver_Type_None;
+    }
+
+    static const struct AP_Param::GroupInfo var_info[];
+
+private:
+    
+    // Parameter interface for CANIfaces
+    class CANIface_Params
+    {
+        friend class AP_CANManager;
+
+    public:
+        CANIface_Params()
+        {
+            AP_Param::setup_object_defaults(this, var_info);
+        }
+
+        static const struct AP_Param::GroupInfo var_info[];
+
+    private:
+        AP_Int8 _driver_number;
+        uint8_t _driver_number_cache;
+        AP_Int32 _bitrate;
+    };
+
+    //Parameter Interface for CANDrivers
+    class CANDriver_Params
+    {
+        friend class AP_CANManager;
+
+    public:
+        CANDriver_Params()
+        {
+            AP_Param::setup_object_defaults(this, var_info);
+        }
+        static const struct AP_Param::GroupInfo var_info[];
+
+    private:
+        AP_Int8 _driver_type;
```

Comment:
No, we don't call interface "DRIVER" in parameters. Infact we call Protocol Drivers "DRIVER":

* We use CAN_P# for Interface settings, and CAN_D# for Protocol Drivers settings. D being clearly for Driver.

* Secondly using the parameter CAN_P#_DRIVER we select which Protocol Driver Number will be used.

* With CAN_D#_PROTOCOL we select the id of PROTOCOL DRIVER to be used.

I have marked a clear distinction of calling protocol_driver "driver" or "protocol_driver", and calling hardware_interface interface. It is clear and well followed everywhere to avoid confusion.
Previously it was a mess of names used interchangeably which was utterly confusing.

Reply:
i see both arguments and i'm happy with either way, as long as there is a comment on the variable declaration saying which parameter this aligns with